spec/models/referral_spec.rb

```diff
@@ -0,0 +1,91 @@
+require 'rails_helper'
+
+RSpec.describe Referral, type: :model do
+  subject {
+    Referral.new(
+      id: "test-id",
+      code: "12356789",
+      person_name: "Test Person",
+      service_provider: "Test Service Prov.ider"
+    )
+  }
+
+  describe "#id" do
+    it "returns an id" do
+      expect(subject.id).to eql("test-id")
+    end
+  end
+
+  describe '#code' do
+    it 'returns a code' do
+      expect(subject.code).to eql("12356789")
+    end
+  end
+
+  describe "#person_name" do
+    it "returns the person's name" do
+      expect(subject.person_name).to eql("Test Person")
+    end
+  end
+
+  describe "#service_provider" do
+    it "returns a service provider" do
+      expect(subject.service_provider).to eql("Test Service Prov.ider")
+    end
+  end
+
+  describe '#service_provider_id' do
+    it "replaces non word characters in #service_provider with underscores" do
+      expect(subject.service_provider_id).to eql("test_service_prov_ider")
+    end
+  end
+
+  describe ".all" do
+    before do
+      allow(Referral).to receive(:worksheet) {
+        OpenStruct.new(rows: [
+          ["Who (optional)", "Service", "Link / code", "Include in public site?"],
+          ["", "Service 1", "Code A", "Y"],
+          ["Bob", "Service 1", "Code B", "Y"],
+          ["Alice", "Service 2", "Code C", ""]
+        ])
+      }
+    end
+    it "returns an array of referrals" do
+      referrals = Referral.all
+
+      expect(referrals).not_to be_empty
+
+      referrals.each do |referral|
```

Comment:
Might be more rubyish to say `referral.map{|x| x.is_a?(Referral)}.all?`, but perhaps also it's less readable. Not sure, but consider. https://ruby-doc.org/core-3.1.2/Enumerable.html#method-i-all-3F